spinalcordtoolbox/centerline/core.py

```diff
@@ -173,11 +175,28 @@ def get_centerline(im_seg, param=ParamCenterline(), verbose=1, remove_temp_files
         # Create an image with the centerline
         im_centerline = im_seg.copy()
         im_centerline.data = np.zeros(im_centerline.data.shape)
-        # Assign value=1 to centerline. Make sure to clip to avoid array overflow.
-        # TODO: check this round and clip-- suspicious
-        im_centerline.data[round_and_clip(x_centerline_fit, clip=[0, im_centerline.data.shape[0]]),
-                           round_and_clip(y_centerline_fit, clip=[0, im_centerline.data.shape[1]]),
-                           z_ref] = 1
+        # Binarized
+        if param.soft == 0:
+            # Assign value=1 to centerline. Make sure to clip to avoid array overflow.
+            im_centerline.data[round_and_clip(x_centerline_fit, clip=[0, im_centerline.data.shape[0]]),
+                               round_and_clip(y_centerline_fit, clip=[0, im_centerline.data.shape[1]]),
+                               z_ref] = 1
+        # Soft (accounting for partial volume effect)
+        else:
+            # Compute integer pixel locations (all together, these form a 2x2 grid around centerline)
+            x_ceil = np.ceil(x_centerline_fit).astype(int)
+            y_ceil = np.ceil(y_centerline_fit).astype(int)
+            x_floor = np.floor(x_centerline_fit).astype(int)
+            y_floor = np.floor(y_centerline_fit).astype(int),
+            # Compute the difference between the centerline and its rounded values
+            x_diff = x_centerline_fit - x_centerline_fit.astype(int)
+            y_diff = y_centerline_fit - y_centerline_fit.astype(int)
+            # Assign soft segmentation values to the 2x2 grid based on the computed differences
+            im_centerline.data[x_ceil, y_ceil, z_ref] = 0.5 * (x_diff + y_diff)
+            im_centerline.data[x_ceil, y_floor, z_ref] = 0.5 * (x_diff + (1 - y_diff))
+            im_centerline.data[x_floor, y_ceil, z_ref] = 0.5 * ((1 - x_diff) + y_diff)
+            im_centerline.data[x_floor, y_floor, z_ref] = 0.5 * ((1 - x_diff) + (1 - y_diff))
```

Comment:
Also, it might be worth writing a test for this? e.g. running `sct_get_centerline` twice (`-soft 0` and `-soft 1`), then verifying that the two centerlines A) overlap and B) the soft centerline adds to `1`.

That way, we can modify this code in the future and verify that it still produces expected output.

Reply:
Conversation superseded by https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4026#discussion_r1096093021.